File: login-provider/android/src/main/java/net/bitburst/plugins/loginprovider/LoginProviderPlugin.java
package net.bitburst.plugins.loginprovider;

import android.content.Intent;

import androidx.activity.result.ActivityResult;
import androidx.annotation.Nullable;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.twitter.sdk.android.core.TwitterAuthConfig;

import net.bitburst.plugins.loginprovider.providers.FacebookProvider;
import net.bitburst.plugins.loginprovider.providers.GoogleProvider;
import net.bitburst.plugins.loginprovider.providers.TwitterProvider;

import java.util.Objects;

@CapacitorPlugin(
    name = "LoginProvider",
    permissions = { @Permission(alias = "internet", strings = {}) },
    requestCodes = { TwitterAuthConfig.DEFAULT_AUTH_REQUEST_CODE }
)
public class LoginProviderPlugin extends Plugin {

    public interface AppStatusChangeListener {
        void onAppStatusChanged(Boolean isActive);
    }

    public static final String LOG_TAG = "LoginProvider ";
    public static final String AUTH_STATE_CHANGE_EVENT = "authStateChange";

    private AppStatusChangeListener statusChangeListener;

    private FacebookProvider facebookProvider;
    private GoogleProvider googleProvider;
    private TwitterProvider twitterProvider = null;

    private JSObject facebookOptions;
    private JSObject googleOptions;
    private JSObject twitterOptions;

    @Override
    public void load() {
        super.load();
        initializeProviders();
    }

    private void initializeProviders() {
        facebookOptions = LoginProviderHelper.convertJSONObject(getConfig().getObject("facebook"));
        googleOptions = LoginProviderHelper.convertJSONObject(getConfig().getObject("google"));
        twitterOptions = LoginProviderHelper.convertJSONObject(getConfig().getObject("twitter"));

        facebookProvider = new FacebookProvider(this, facebookOptions);
        googleProvider = new GoogleProvider(this, googleOptions);
        twitterProvider = getTwitterProviderInstance();
    }

    @PluginMethod
    public void loginWithProvider(PluginCall call) {
        if (call == null) return;
        if (!call.hasOption("provider")) call.reject("provider name missing");

        try {
            String providerName = call.getString("provider");

            //   call.setKeepAlive(true);
            switch (Objects.requireNonNull(providerName)) {
                case "APPLE":
                    call.unimplemented();
                    break;
                case "GOOGLE":
                    googleProvider.login(call);
                    break;
                case "FACEBOOK":
                    facebookProvider.login(call);
                    break;
                case "TWITTER":
                    getTwitterProviderInstance().login(call);
                    break;
            }
        } catch (Exception ex) {
            call.reject(ex.getLocalizedMessage());
        }
    }

    private TwitterProvider getTwitterProviderInstance() {
        if (twitterProvider == null) {
            twitterProvider = new TwitterProvider(this, twitterOptions);
        }

        return twitterProvider;
    }

    @PluginMethod
    public void loginWithApple(PluginCall call) {
        call.unimplemented();
    }

    @PluginMethod
    public void loginWithFacebook(PluginCall call) {
        call.unimplemented();
    }

    @PluginMethod
    public void loginWithGoogle(PluginCall call) {
        call.unimplemented();
    }

    @PluginMethod
    public void loginWithTwitter(PluginCall call) {
        call.unimplemented();
    }

    @PluginMethod
    public void logoutFromProvider(PluginCall call) {
        if (call == null) return;
        if (!call.hasOption("provider")) call.reject("provider name missing");

        try {
            String providerName = call.getString("provider");
            switch (Objects.requireNonNull(providerName)) {
                case "APPLE":
                    call.unimplemented();
                    break;
                case "GOOGLE":
                    googleProvider.logout();
                    break;
                case "FACEBOOK":
                    facebookProvider.logout(call);
                    break;
                case "TWITTER":
                    twitterProvider.logout(call);
                    break;
            }
        } catch (Exception ex) {
            call.reject(ex.getLocalizedMessage());
        }
    }

    @PluginMethod
    public void addListener(PluginCall call) {
        call.unimplemented();
    }

    @PluginMethod
    public void removeAllListeners(PluginCall call) {
        call.unimplemented();
    }

    @ActivityCallback
    protected void facebookLoginRequest(PluginCall call, ActivityResult result) {
        facebookProvider.handleLoginRequest(call, result);
    }

    @ActivityCallback
    protected void googleLoginRequest(PluginCall call, ActivityResult result) {
        googleProvider.handleLoginRequest(call, result);
    }

    @ActivityCallback
    protected void twitterLoginResult(PluginCall call, ActivityResult result) {}

    @SuppressWarnings("deprecation")
    @Override
    public void handleOnActivityResult(int requestCode, int resultCode, Intent data) {
        super.handleOnActivityResult(requestCode, resultCode, data);
        getTwitterProviderInstance().handleLoginResult(TwitterAuthConfig.DEFAULT_AUTH_REQUEST_CODE, resultCode, data);
    }

    public void notifyListeners(String eventName, JSObject data) {
        super.notifyListeners(eventName, data);
    }

    public void setAppStatusChangeListener(@Nullable AppStatusChangeListener listener) {
        statusChangeListener = listener;
    }

    @Nullable
    public AppStatusChangeListener getAppStatusChangeListener() {
        return statusChangeListener;
    }
}

File: login-provider/ios/Plugin/LoginProviderPlugin.swift
import Foundation
import Capacitor

typealias JSObject = [String: Any]
typealias JSArray = [JSObject]
typealias ProvidersMap = [String: ProviderHandler]

@objc(LoginProviderPlugin)
public class LoginProviderPlugin: CAPPlugin {

    var appleProvider: AppleProvider?
    var facebookProvider: FacebookProvider?
    var googleProvider: GoogleProvider?
    var twitterProvider: TwitterProvider?

    var providers: ProvidersMap = [:]
    var callbackId: String?

    var mOptions: JSObject?

    override public func load() {
        let appleOptions = getConfigValue("apple")
        let facebookOptions = getConfigValue("facebook")
        let googleOptions = getConfigValue("google")
        let twitterOptions = getConfigValue("twitter")

        self.providers["APPLE"] = AppleProvider()
        self.providers["APPLE"]?.initialize(plugin: self, options: appleOptions as! JSObject)

        self.providers["FACEBOOK"] = FacebookProvider()
        self.providers["FACEBOOK"]?.initialize(plugin: self, options: facebookOptions as! JSObject)

        self.providers["GOOGLE"] = GoogleProvider()
        self.providers["GOOGLE"]?.initialize(plugin: self, options: googleOptions as! JSObject)

        self.providers["TWITTER"] = TwitterProvider()
        self.providers["TWITTER"]?.initialize(plugin: self, options: twitterOptions as! JSObject)
    }

    @objc func loginWithProvider(_ call: CAPPluginCall) {
        guard let providerInstance: ProviderHandler = self.getProvider(call: call) else {
            // call.reject inside getProvider
            return
        }

        guard let callbackId = call.callbackId else {
            call.reject("The call has no callbackId")
            return
        }

        self.callbackId = callbackId
        call.keepAlive = true

        DispatchQueue.main.async {
            if providerInstance.isAuthenticated() {
                self.buildResult(credential: nil)
                return
            }

            providerInstance.login(call: call)
        }
    }

    @objc func logoutFromProvider(call: CAPPluginCall) {
        do {
            for provider in self.providers.values {
                try provider.logout(call: call)
            }
            call.resolve()
        } catch let logoutError as NSError {
            NSLog("logout error: %@", logoutError)
            call.reject("logout error: \(logoutError)")
        }
    }

    func getProvider(call: CAPPluginCall) -> ProviderHandler? {
        guard let providerId = call.getString("provider") else {
            call.reject("provider name is required")
            return nil
        }

        guard let theProvider = self.providers[providerId] else {
            call.reject("provider not supported")
            return nil
        }

        return theProvider
    }

    func buildResult(credential: ProviderResponsePayload?) {
        guard let callbackId = self.callbackId else {
            NSLog("missing plugin callback id")
            return
        }

        guard let call = self.bridge?.savedCall(withID: callbackId) else {
            NSLog("missing plugin callback id")
            return
        }

        let jsResult: PluginCallResultData = [
            "callbackId": callbackId,
            "providerId": call.getString("provider") ?? ""
        ]

        guard let provider: ProviderHandler = self.getProvider(call: call) else {
            return
        }

        call.resolve(provider.fillResult(data: jsResult))
    }

    @objc func loginWithApple(_ call: CAPPluginCall) {
        call.unimplemented()
    }

    @objc func loginWithFacebook(_ call: CAPPluginCall) {
        call.unimplemented()
    }

    @objc func loginWithGoogle(_ call: CAPPluginCall) {
        call.unimplemented()
    }

    @objc func loginWithTwitter(_ call: CAPPluginCall) {
        call.unimplemented()
    }

    @objc override public func addListener(_ call: CAPPluginCall) {
        call.unimplemented()
    }

    @objc override public func removeAllListeners(_ call: CAPPluginCall) {
        call.unimplemented()
    }
}
